### ISSUE TITLE
Regression in `jcouyang/dhall-generic` for implicit search 

### ISSUE THREAD

**Issue by WojciechMazur** [scala/scala3#20298]
Based on OpenCB failures in `jcouyang/dhall-generic` - [build logs](https://github.com/VirtusLab/community-build3/actions/runs/8855242584/job/24320259671)

## Compiler version
Last good release: 3.5.0-RC1-bin-20240406-521ce95-NIGHTLY
First bad release: 3.5.0-RC1-bin-20240408-1e8a653-NIGHTLY

The compilation errors when building compiler prevented gathering the exact commit that coused regression it can be any of: 
d2a6392131001af941b0614c9e6d812e8c761965
1dc5b995cfb84dd3e2673041d6e954ed6043900e
321d0d2cdee757fe311c9959817794acaad0fbdb
1e8a653cce58ddb038c04a906448db92172828dd

The [project-wise bisect](https://github.com/VirtusLab/community-build3/actions/runs/8885223691/job/24396091199) nominated 90c3fbde37a70fdd366ed25d2082e05ddf24ac3e as the cause of the regression

## Minimized code
Needs to be defined as 2 compilation units
I was able to reproducer `Decoder[Empty]` in single file, but the other cases required definitions and usage to be split in previous Scala versions. 

```Scala
// 1_defns.scala
import scala.compiletime.*
import scala.deriving.*

sealed trait Decoder[T]
object Decoder:
  inline given summonEmptyTuple[H]: Tuple.Map[EmptyTuple.type, Decoder] = EmptyTuple
  inline given summonTuple[H, T <: Tuple](using
      hd: Decoder[H],
      td: Tuple.Map[T, Decoder]
  ): Tuple.Map[H *: T, Decoder] = hd *: td
  inline given derived[T](using
      m: Mirror.Of[T],
      d: Tuple.Map[m.MirroredElemTypes, Decoder]
  ): Decoder[T] = ???
  given Decoder[Double] = ???
```

```scala
// 2_usage.scala
case class Empty()
object Empty {
  val fails = Decoder.derived[Empty]
}

enum Shape:
  case Rectangle(width: Double, height: Double)
  case Circle(radius: Double)

object Shape:
  val fails1 = Decoder.derived[Rectangle]
  val fails2 = Decoder.derived[Shape]
```

## Output

```scala
-- [E172] Type Error: /Users/wmazur/projects/sandbox/repro.test.scala:4:39 -----
4 |  val instance = Decoder.derived[Empty]
  |                                       ^
  |No given instance of type Tuple.Map[m.MirroredElemTypes, Decoder] was found for parameter d of given instance derived in object Decoder
  |
  |Note: a match type could not be fully reduced:
  |
  |  trying to reduce  Tuple.Map[m.MirroredElemTypes, Decoder]
  |  failed since selector m.MirroredElemTypes
  |  does not match  case EmptyTuple => EmptyTuple
  |  and cannot be shown to be disjoint from it either.
  |  Therefore, reduction cannot advance to the remaining case
  |
  |    case h *: t => Decoder[h] *: Tuple.Map[t, Decoder]
-- [E172] Type Error: /Users/wmazur/projects/sandbox/repro.test.scala:12:38 ----
12 |  val rec = Decoder.derived[Rectangle]
   |                                      ^
   |No given instance of type Tuple.Map[m.MirroredElemTypes, Decoder] was found for parameter d of given instance derived in object Decoder.
   |I found:
   |
   |    Decoder.summonTuple[Double, Double *: EmptyTuple.type](
   |      Decoder.given_Decoder_Double,
   |      Decoder.summonTuple[Double, EmptyTuple.type](Decoder.given_Decoder_Double,
   |        /* missing */summon[Tuple.Map[EmptyTuple.type, Decoder]])
   |    )
   |
   |But no implicit values were found that match type Tuple.Map[EmptyTuple.type, Decoder].
   |
   |Note: a match type could not be fully reduced:
   |
   |  trying to reduce  Tuple.Map[m.MirroredElemTypes, Decoder]
   |  failed since selector m.MirroredElemTypes
   |  does not match  case EmptyTuple => EmptyTuple
   |  and cannot be shown to be disjoint from it either.
   |  Therefore, reduction cannot advance to the remaining case
   |
   |    case h *: t => Decoder[h] *: Tuple.Map[t, Decoder]
3 errors found
```

## Expectation
Should compile

**Comment by EugeneFlesselle** [comment]
I believe this _is_ an expected error. Or at least, we intended for it to be an error in #20077. 
The `i15183` test looks to be a very similar minimisation as above, it was [changed](https://github.com/scala/scala3/pull/20077/files#diff-4ac552f3bea7d4adc48de6e328a2428804fbb00c99e267b4079473a76d7dabcc) in the PR to account for the fix in determining implicit scope. 

TLDR the regression is necessary for the implicit scope to be the same for joint vs separate compilation.
I believe it would be fixed here too by adding an `import Decoder.given` in `2_usage.scala`.
